// promptfoo/dist/src/redteam/plugins/harmful/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHarmfulAssertions = getHarmfulAssertions;
exports.createTestCase = createTestCase;
const envars_1 = require("../../../envars");
const constants_1 = require("../../constants");
function getHarmfulAssertions(harmCategory) {
    const assertionType = harmCategory === 'harmful:privacy'
        ? 'promptfoo:redteam:harmful:privacy'
        : 'promptfoo:redteam:harmful';
    const assertions = [{ metric: 'Harmful', type: assertionType }];
    if ((0, envars_1.getEnvBool)('PROMPTFOO_DISABLE_REDTEAM_MODERATION')) {
        return assertions;
    }
    if ((0, envars_1.getEnvString)('REPLICATE_API_KEY') || (0, envars_1.getEnvString)('REPLICATE_API_TOKEN')) {
        assertions.push({
            metric: 'Harmful',
            type: 'moderation',
            value: constants_1.LLAMA_GUARD_ENABLED_CATEGORIES,
            provider: constants_1.LLAMA_GUARD_REPLICATE_PROVIDER,
        });
    }
    else if ((0, envars_1.getEnvString)('OPENAI_API_KEY')) {
        assertions.push({ metric: 'Harmful', type: 'moderation' });
    }
    return assertions;
}
function createTestCase(injectVar, output, harmCategory) {
    const harmCategoryLabel = constants_1.HARM_PLUGINS[harmCategory] || harmCategory;
    return {
        vars: {
            [injectVar]: output.trim(),
            harmCategory: harmCategoryLabel,
        },
        metadata: { harmCategory: harmCategoryLabel },
        assert: getHarmfulAssertions(harmCategory),
    };
}
