// promptfoo/dist/src/util/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeOutput = writeOutput;
exports.writeMultipleOutputs = writeMultipleOutputs;
exports.readOutput = readOutput;
exports.getLatestResultsPath = getLatestResultsPath;
exports.writeResultsToDatabase = writeResultsToDatabase;
exports.listPreviousResults = listPreviousResults;
exports.listPreviousResultFilenames_fileSystem = listPreviousResultFilenames_fileSystem;
exports.listPreviousResults_fileSystem = listPreviousResults_fileSystem;
exports.filenameToDate = filenameToDate;
exports.dateToFilename = dateToFilename;
exports.readResult_fileSystem = readResult_fileSystem;
exports.readResult = readResult;
exports.updateResult = updateResult;
exports.getLatestEval = getLatestEval;
exports.getPromptsWithPredicate = getPromptsWithPredicate;
exports.getPromptsForTestCasesHash = getPromptsForTestCasesHash;
exports.getPromptsForTestCases = getPromptsForTestCases;
exports.getTestCasesWithPredicate = getTestCasesWithPredicate;
exports.getPrompts = getPrompts;
exports.getTestCases = getTestCases;
exports.getPromptFromHash = getPromptFromHash;
exports.getDatasetFromHash = getDatasetFromHash;
exports.getEvalsWithPredicate = getEvalsWithPredicate;
exports.getEvals = getEvals;
exports.getEvalFromId = getEvalFromId;
exports.deleteEval = deleteEval;
exports.deleteAllEvals = deleteAllEvals;
exports.readFilters = readFilters;
exports.printBorder = printBorder;
exports.setupEnv = setupEnv;
exports.getStandaloneEvals = getStandaloneEvals;
exports.providerToIdentifier = providerToIdentifier;
exports.varsMatch = varsMatch;
exports.resultIsForTestCase = resultIsForTestCase;
exports.renderVarsInObject = renderVarsInObject;
exports.parsePathOrGlob = parsePathOrGlob;
exports.maybeLoadFromExternalFile = maybeLoadFromExternalFile;
exports.isRunningUnderNpx = isRunningUnderNpx;
const sync_1 = require("csv-parse/sync");
const sync_2 = require("csv-stringify/sync");
const dedent_1 = __importDefault(require("dedent"));
const dotenv_1 = __importDefault(require("dotenv"));
const drizzle_orm_1 = require("drizzle-orm");
const fast_deep_equal_1 = __importDefault(require("fast-deep-equal"));
const fs = __importStar(require("fs"));
const glob_1 = require("glob");
const js_yaml_1 = __importDefault(require("js-yaml"));
const node_cache_1 = __importDefault(require("node-cache"));
const nunjucks_1 = __importDefault(require("nunjucks"));
const path = __importStar(require("path"));
const cliState_1 = __importDefault(require("../cliState"));
const constants_1 = require("../constants");
const database_1 = require("../database");
const tables_1 = require("../database/tables");
const envars_1 = require("../envars");
const esm_1 = require("../esm");
const accounts_1 = require("../globalConfig/accounts");
const googleSheets_1 = require("../googleSheets");
const logger_1 = __importDefault(require("../logger"));
const eval_1 = __importStar(require("../models/eval"));
const prompt_1 = require("../models/prompt");
const types_1 = require("../types");
const invariant_1 = __importDefault(require("../util/invariant"));
const manage_1 = require("./config/manage");
const createHash_1 = require("./createHash");
const exportToFile_1 = require("./exportToFile");
const file_1 = require("./file");
const templates_1 = require("./templates");
const DEFAULT_QUERY_LIMIT = 100;
const outputToSimpleString = (output) => {
    const passFailText = output.pass
        ? '[PASS]'
        : output.failureReason === types_1.ResultFailureReason.ASSERT
            ? '[FAIL]'
            : '[ERROR]';
    const namedScoresText = Object.entries(output.namedScores)
        .map(([name, value]) => `${name}: ${value?.toFixed(2)}`)
        .join(', ');
    const scoreText = namedScoresText.length > 0
        ? `(${output.score?.toFixed(2)}, ${namedScoresText})`
        : `(${output.score?.toFixed(2)})`;
    const gradingResultText = output.gradingResult
        ? `${output.pass ? 'Pass' : 'Fail'} Reason: ${output.gradingResult.reason}`
        : '';
    return (0, dedent_1.default) `
      ${passFailText} ${scoreText}

      ${output.text}

      ${gradingResultText}
    `.trim();
};
async function writeOutput(outputPath, evalRecord, shareableUrl) {
    if (outputPath.match(/^https:\/\/docs\.google\.com\/spreadsheets\//)) {
        const table = await evalRecord.getTable();
        (0, invariant_1.default)(table, 'Table is required');
        const rows = table.body.map((row) => {
            const csvRow = {};
            table.head.vars.forEach((varName, index) => {
                csvRow[varName] = row.vars[index];
            });
            table.head.prompts.forEach((prompt, index) => {
                csvRow[prompt.label] = outputToSimpleString(row.outputs[index]);
            });
            return csvRow;
        });
        logger_1.default.info(`Writing ${rows.length} rows to Google Sheets...`);
        await (0, googleSheets_1.writeCsvToGoogleSheet)(rows, outputPath);
        return;
    }
    const { data: outputExtension } = types_1.OutputFileExtension.safeParse(path.extname(outputPath).slice(1).toLowerCase());
    (0, invariant_1.default)(outputExtension, `Unsupported output file format ${outputExtension}. Please use one of: ${types_1.OutputFileExtension.options.join(', ')}.`);
    // Ensure the directory exists
    const outputDir = path.dirname(outputPath);
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    if (outputExtension === 'csv') {
        // Write headers first
        const headers = (0, exportToFile_1.getHeaderForTable)(evalRecord);
        const headerCsv = (0, sync_2.stringify)([
            [...headers.vars, ...headers.prompts.map((prompt) => `[${prompt.provider}] ${prompt.label}`)],
        ]);
        fs.writeFileSync(outputPath, headerCsv);
        // Write body rows in batches
        for await (const batchResults of evalRecord.fetchResultsBatched()) {
            // we need split the batch into rows by testIdx
            const tableRows = {};
            for (const result of batchResults) {
                if (!(result.testIdx in tableRows)) {
                    tableRows[result.testIdx] = [];
                }
                tableRows[result.testIdx].push(result);
            }
            const batchCsv = (0, sync_2.stringify)(Object.values(tableRows).map((results) => {
                const row = (0, exportToFile_1.convertTestResultsToTableRow)(results, headers.vars);
                return [...row.vars, ...row.outputs.map(outputToSimpleString)];
            }));
            console.log({ batchCsv });
            fs.appendFileSync(outputPath, batchCsv);
        }
    }
    else if (outputExtension === 'json') {
        const summary = await evalRecord.toEvaluateSummary();
        fs.writeFileSync(outputPath, JSON.stringify({
            evalId: evalRecord.id,
            results: summary,
            config: evalRecord.config,
            shareableUrl,
        }, null, 2));
    }
    else if (outputExtension === 'yaml' || outputExtension === 'yml' || outputExtension === 'txt') {
        const summary = await evalRecord.toEvaluateSummary();
        fs.writeFileSync(outputPath, js_yaml_1.default.dump({
            evalId: evalRecord.id,
            results: summary,
            config: evalRecord.config,
            shareableUrl,
        }));
    }
    else if (outputExtension === 'html') {
        const table = await evalRecord.getTable();
        (0, invariant_1.default)(table, 'Table is required');
        const summary = await evalRecord.toEvaluateSummary();
        const template = fs.readFileSync(`${(0, esm_1.getDirectory)()}/tableOutput.html`, 'utf-8');
        const htmlTable = [
            [
                ...table.head.vars,
                ...table.head.prompts.map((prompt) => `[${prompt.provider}] ${prompt.label}`),
            ],
            ...table.body.map((row) => [...row.vars, ...row.outputs.map(outputToSimpleString)]),
        ];
        const htmlOutput = (0, templates_1.getNunjucksEngine)().renderString(template, {
            config: evalRecord.config,
            table: htmlTable,
            results: summary,
        });
        fs.writeFileSync(outputPath, htmlOutput);
    }
    else if (outputExtension === 'jsonl') {
        for await (const batchResults of evalRecord.fetchResultsBatched()) {
            const text = batchResults.map((result) => JSON.stringify(result)).join('\n');
            fs.appendFileSync(outputPath, text);
        }
    }
}
async function writeMultipleOutputs(outputPaths, evalRecord, shareableUrl) {
    await Promise.all(outputPaths.map((outputPath) => writeOutput(outputPath, evalRecord, shareableUrl)));
}
async function readOutput(outputPath) {
    const ext = path.parse(outputPath).ext.slice(1);
    switch (ext) {
        case 'json':
            return JSON.parse(fs.readFileSync(outputPath, 'utf-8'));
        default:
            throw new Error(`Unsupported output file format: ${ext} currently only supports json`);
    }
}
/**
 * TODO(ian): Remove this
 * @deprecated Use readLatestResults directly instead.
 */
function getLatestResultsPath() {
    return path.join((0, manage_1.getConfigDirectoryPath)(), 'output', 'latest.json');
}
async function writeResultsToDatabase(results, config, createdAt = new Date()) {
    createdAt = createdAt || (results.timestamp ? new Date(results.timestamp) : new Date());
    const evalId = (0, eval_1.createEvalId)(createdAt);
    const db = (0, database_1.getDb)();
    const promises = [];
    promises.push(db
        .insert(tables_1.evalsTable)
        .values({
        id: evalId,
        createdAt: createdAt.getTime(),
        author: (0, accounts_1.getAuthor)(),
        description: config.description,
        config,
        results,
    })
        .onConflictDoNothing()
        .run());
    logger_1.default.debug(`Inserting eval ${evalId}`);
    // Record prompt relation
    (0, invariant_1.default)(results.table, 'Table is required');
    for (const prompt of results.table.head.prompts) {
        const label = prompt.label || prompt.display || prompt.raw;
        const promptId = (0, prompt_1.generateIdFromPrompt)(prompt);
        promises.push(db
            .insert(tables_1.promptsTable)
            .values({
            id: promptId,
            prompt: label,
        })
            .onConflictDoNothing()
            .run());
        promises.push(db
            .insert(tables_1.evalsToPromptsTable)
            .values({
            evalId,
            promptId,
        })
            .onConflictDoNothing()
            .run());
        logger_1.default.debug(`Inserting prompt ${promptId}`);
    }
    // Record dataset relation
    const datasetId = (0, createHash_1.sha256)(JSON.stringify(config.tests || []));
    promises.push(db
        .insert(tables_1.datasetsTable)
        .values({
        id: datasetId,
        tests: config.tests,
    })
        .onConflictDoNothing()
        .run());
    promises.push(db
        .insert(tables_1.evalsToDatasetsTable)
        .values({
        evalId,
        datasetId,
    })
        .onConflictDoNothing()
        .run());
    logger_1.default.debug(`Inserting dataset ${datasetId}`);
    // Record tags
    if (config.tags) {
        for (const [tagKey, tagValue] of Object.entries(config.tags)) {
            const tagId = (0, createHash_1.sha256)(`${tagKey}:${tagValue}`);
            promises.push(db
                .insert(tables_1.tagsTable)
                .values({
                id: tagId,
                name: tagKey,
                value: tagValue,
            })
                .onConflictDoNothing()
                .run());
            promises.push(db
                .insert(tables_1.evalsToTagsTable)
                .values({
                evalId,
                tagId,
            })
                .onConflictDoNothing()
                .run());
            logger_1.default.debug(`Inserting tag ${tagId}`);
        }
    }
    logger_1.default.debug(`Awaiting ${promises.length} promises to database...`);
    await Promise.all(promises);
    return evalId;
}
/**
 *
 * @returns Last n evals in descending order.
 */
async function listPreviousResults(limit = DEFAULT_QUERY_LIMIT, filterDescription, datasetId) {
    const db = (0, database_1.getDb)();
    const startTime = performance.now();
    const query = db
        .select({
        evalId: tables_1.evalsTable.id,
        createdAt: tables_1.evalsTable.createdAt,
        description: tables_1.evalsTable.description,
        numTests: (0, drizzle_orm_1.sql) `json_array_length(${tables_1.evalsTable.results}->'table'->'body')`,
        datasetId: tables_1.evalsToDatasetsTable.datasetId,
        isRedteam: (0, drizzle_orm_1.sql) `json_type(${tables_1.evalsTable.config}, '$.redteam') IS NOT NULL`,
    })
        .from(tables_1.evalsTable)
        .leftJoin(tables_1.evalsToDatasetsTable, (0, drizzle_orm_1.eq)(tables_1.evalsTable.id, tables_1.evalsToDatasetsTable.evalId))
        .where((0, drizzle_orm_1.and)(datasetId ? (0, drizzle_orm_1.eq)(tables_1.evalsToDatasetsTable.datasetId, datasetId) : undefined, filterDescription ? (0, drizzle_orm_1.like)(tables_1.evalsTable.description, `%${filterDescription}%`) : undefined, (0, drizzle_orm_1.not)((0, drizzle_orm_1.eq)(tables_1.evalsTable.results, {}))));
    const results = query.orderBy((0, drizzle_orm_1.desc)(tables_1.evalsTable.createdAt)).limit(limit).all();
    const mappedResults = results.map((result) => ({
        evalId: result.evalId,
        createdAt: result.createdAt,
        description: result.description,
        numTests: result.numTests,
        datasetId: result.datasetId,
        isRedteam: result.isRedteam,
    }));
    const endTime = performance.now();
    const executionTime = endTime - startTime;
    const evalResults = await (0, eval_1.getSummaryOfLatestEvals)(undefined, filterDescription, datasetId);
    logger_1.default.debug(`listPreviousResults execution time: ${executionTime.toFixed(2)}ms`);
    const combinedResults = [...evalResults, ...mappedResults];
    return combinedResults;
}
/**
 * @deprecated Used only for migration to sqlite
 */
function listPreviousResultFilenames_fileSystem() {
    const directory = path.join((0, manage_1.getConfigDirectoryPath)(), 'output');
    if (!fs.existsSync(directory)) {
        return [];
    }
    const files = fs.readdirSync(directory);
    const resultsFiles = files.filter((file) => file.startsWith('eval-') && file.endsWith('.json'));
    return resultsFiles.sort((a, b) => {
        const statA = fs.statSync(path.join(directory, a));
        const statB = fs.statSync(path.join(directory, b));
        return statA.birthtime.getTime() - statB.birthtime.getTime(); // sort in ascending order
    });
}
const resultsCache = {};
/**
 * @deprecated Used only for migration to sqlite
 */
function listPreviousResults_fileSystem() {
    const directory = path.join((0, manage_1.getConfigDirectoryPath)(), 'output');
    if (!fs.existsSync(directory)) {
        return [];
    }
    const sortedFiles = listPreviousResultFilenames_fileSystem();
    return sortedFiles.map((fileName) => {
        if (!resultsCache[fileName]) {
            try {
                const fileContents = fs.readFileSync(path.join(directory, fileName), 'utf8');
                const data = js_yaml_1.default.load(fileContents);
                resultsCache[fileName] = data;
            }
            catch (error) {
                logger_1.default.warn(`Failed to read results from ${fileName}:\n${error}`);
            }
        }
        return {
            fileName,
            description: resultsCache[fileName]?.config.description,
        };
    });
}
function filenameToDate(filename) {
    const dateString = filename.slice('eval-'.length, filename.length - '.json'.length);
    // Replace hyphens with colons where necessary (Windows compatibility).
    const dateParts = dateString.split('T');
    const timePart = dateParts[1].replace(/-/g, ':');
    const formattedDateString = `${dateParts[0]}T${timePart}`;
    const date = new Date(formattedDateString);
    return date;
    /*
    return date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      timeZoneName: 'short',
    });
    */
}
function dateToFilename(date) {
    return `eval-${date.toISOString().replace(/:/g, '-')}.json`;
}
/**
 * @deprecated Used only for migration to sqlite
 */
function readResult_fileSystem(name) {
    const resultsDirectory = path.join((0, manage_1.getConfigDirectoryPath)(), 'output');
    const resultsPath = path.join(resultsDirectory, name);
    try {
        const result = JSON.parse(fs.readFileSync(fs.realpathSync(resultsPath), 'utf-8'));
        const createdAt = filenameToDate(name);
        return {
            id: (0, createHash_1.sha256)(JSON.stringify(result.config)),
            result,
            createdAt,
        };
    }
    catch (err) {
        logger_1.default.error(`Failed to read results from ${resultsPath}:\n${err}`);
    }
}
async function readResult(id) {
    try {
        const eval_ = await eval_1.default.findById(id);
        (0, invariant_1.default)(eval_, `Eval with ID ${id} not found.`);
        return {
            id,
            result: await eval_.toResultsFile(),
            createdAt: new Date(eval_.createdAt),
        };
    }
    catch (err) {
        logger_1.default.error(`Failed to read result with ID ${id} from database:\n${err}`);
    }
}
async function updateResult(id, newConfig, newTable) {
    try {
        // Fetch the existing eval data from the database
        const existingEval = await eval_1.default.findById(id);
        if (!existingEval) {
            logger_1.default.error(`Eval with ID ${id} not found.`);
            return;
        }
        if (newConfig) {
            existingEval.config = newConfig;
        }
        if (newTable) {
            existingEval.setTable(newTable);
        }
        await existingEval.save();
        logger_1.default.info(`Updated eval with ID ${id}`);
    }
    catch (err) {
        logger_1.default.error(`Failed to update eval with ID ${id}:\n${err}`);
    }
}
async function getLatestEval(filterDescription) {
    const eval_ = await eval_1.default.latest();
    return await eval_?.toResultsFile();
}
async function getPromptsWithPredicate(predicate, limit) {
    // TODO(ian): Make this use a proper database query
    const evals_ = await eval_1.default.getMany(limit);
    const groupedPrompts = {};
    for (const eval_ of evals_) {
        const createdAt = new Date(eval_.createdAt).toISOString();
        const resultWrapper = await eval_.toResultsFile();
        if (predicate(resultWrapper)) {
            for (const prompt of eval_.getPrompts()) {
                const promptId = (0, createHash_1.sha256)(prompt.raw);
                const datasetId = resultWrapper.config.tests
                    ? (0, createHash_1.sha256)(JSON.stringify(resultWrapper.config.tests))
                    : '-';
                if (promptId in groupedPrompts) {
                    groupedPrompts[promptId].recentEvalDate = new Date(Math.max(groupedPrompts[promptId].recentEvalDate.getTime(), new Date(createdAt).getTime()));
                    groupedPrompts[promptId].count += 1;
                    groupedPrompts[promptId].evals.push({
                        id: eval_.id,
                        datasetId,
                        metrics: prompt.metrics,
                    });
                }
                else {
                    groupedPrompts[promptId] = {
                        count: 1,
                        id: promptId,
                        prompt,
                        recentEvalDate: new Date(createdAt),
                        recentEvalId: eval_.id,
                        evals: [
                            {
                                id: eval_.id,
                                datasetId,
                                metrics: prompt.metrics,
                            },
                        ],
                    };
                }
            }
        }
    }
    return Object.values(groupedPrompts);
}
function getPromptsForTestCasesHash(testCasesSha256, limit = DEFAULT_QUERY_LIMIT) {
    return getPromptsWithPredicate((result) => {
        const testsJson = JSON.stringify(result.config.tests);
        const hash = (0, createHash_1.sha256)(testsJson);
        return hash === testCasesSha256;
    }, limit);
}
function getPromptsForTestCases(testCases) {
    const testCasesJson = JSON.stringify(testCases);
    const testCasesSha256 = (0, createHash_1.sha256)(testCasesJson);
    return getPromptsForTestCasesHash(testCasesSha256);
}
async function getTestCasesWithPredicate(predicate, limit) {
    const evals_ = await eval_1.default.getMany(limit);
    const groupedTestCases = {};
    for (const eval_ of evals_) {
        const createdAt = new Date(eval_.createdAt).toISOString();
        const resultWrapper = await eval_.toResultsFile();
        const testCases = resultWrapper.config.tests;
        if (testCases && predicate(resultWrapper)) {
            const evalId = eval_.id;
            const datasetId = (0, createHash_1.sha256)(JSON.stringify(testCases));
            if (datasetId in groupedTestCases) {
                groupedTestCases[datasetId].recentEvalDate = new Date(Math.max(groupedTestCases[datasetId].recentEvalDate.getTime(), eval_.createdAt));
                groupedTestCases[datasetId].count += 1;
                const newPrompts = eval_.getPrompts().map((prompt) => ({
                    id: (0, createHash_1.sha256)(prompt.raw),
                    prompt,
                    evalId,
                }));
                const promptsById = {};
                for (const prompt of groupedTestCases[datasetId].prompts.concat(newPrompts)) {
                    if (!(prompt.id in promptsById)) {
                        promptsById[prompt.id] = prompt;
                    }
                }
                groupedTestCases[datasetId].prompts = Object.values(promptsById);
            }
            else {
                const newPrompts = eval_.getPrompts().map((prompt) => ({
                    id: (0, createHash_1.sha256)(prompt.raw),
                    prompt,
                    evalId,
                }));
                const promptsById = {};
                for (const prompt of newPrompts) {
                    if (!(prompt.id in promptsById)) {
                        promptsById[prompt.id] = prompt;
                    }
                }
                groupedTestCases[datasetId] = {
                    id: datasetId,
                    count: 1,
                    testCases,
                    recentEvalDate: new Date(createdAt),
                    recentEvalId: evalId,
                    prompts: Object.values(promptsById),
                };
            }
        }
    }
    return Object.values(groupedTestCases);
}
function getPrompts(limit = DEFAULT_QUERY_LIMIT) {
    return getPromptsWithPredicate(() => true, limit);
}
async function getTestCases(limit = DEFAULT_QUERY_LIMIT) {
    return getTestCasesWithPredicate(() => true, limit);
}
async function getPromptFromHash(hash) {
    const prompts = await getPrompts();
    for (const prompt of prompts) {
        if (prompt.id.startsWith(hash)) {
            return prompt;
        }
    }
    return undefined;
}
async function getDatasetFromHash(hash) {
    const datasets = await getTestCases();
    for (const dataset of datasets) {
        if (dataset.id.startsWith(hash)) {
            return dataset;
        }
    }
    return undefined;
}
async function getEvalsWithPredicate(predicate, limit) {
    const db = (0, database_1.getDb)();
    const evals_ = await db
        .select({
        id: tables_1.evalsTable.id,
        createdAt: tables_1.evalsTable.createdAt,
        author: tables_1.evalsTable.author,
        results: tables_1.evalsTable.results,
        config: tables_1.evalsTable.config,
        description: tables_1.evalsTable.description,
    })
        .from(tables_1.evalsTable)
        .orderBy((0, drizzle_orm_1.desc)(tables_1.evalsTable.createdAt))
        .limit(limit)
        .all();
    const ret = [];
    for (const eval_ of evals_) {
        const createdAt = new Date(eval_.createdAt).toISOString();
        const resultWrapper = {
            version: 3,
            createdAt,
            author: eval_.author,
            // @ts-ignore
            results: eval_.results,
            config: eval_.config,
        };
        if (predicate(resultWrapper)) {
            const evalId = eval_.id;
            ret.push({
                id: evalId,
                date: new Date(eval_.createdAt),
                config: eval_.config,
                // @ts-ignore
                results: eval_.results,
                description: eval_.description || undefined,
            });
        }
    }
    return ret;
}
async function getEvals(limit = DEFAULT_QUERY_LIMIT) {
    return getEvalsWithPredicate(() => true, limit);
}
async function getEvalFromId(hash) {
    const evals_ = await getEvals();
    for (const eval_ of evals_) {
        if (eval_.id.startsWith(hash)) {
            return eval_;
        }
    }
    return undefined;
}
async function deleteEval(evalId) {
    const db = (0, database_1.getDb)();
    await db.transaction(async () => {
        // We need to clean up foreign keys first. We don't have onDelete: 'cascade' set on all these relationships.
        await db.delete(tables_1.evalsToPromptsTable).where((0, drizzle_orm_1.eq)(tables_1.evalsToPromptsTable.evalId, evalId)).run();
        await db.delete(tables_1.evalsToDatasetsTable).where((0, drizzle_orm_1.eq)(tables_1.evalsToDatasetsTable.evalId, evalId)).run();
        await db.delete(tables_1.evalsToTagsTable).where((0, drizzle_orm_1.eq)(tables_1.evalsToTagsTable.evalId, evalId)).run();
        await db.delete(tables_1.evalResultsTable).where((0, drizzle_orm_1.eq)(tables_1.evalResultsTable.evalId, evalId)).run();
        // Finally, delete the eval record
        const deletedIds = await db.delete(tables_1.evalsTable).where((0, drizzle_orm_1.eq)(tables_1.evalsTable.id, evalId)).run();
        if (deletedIds.changes === 0) {
            throw new Error(`Eval with ID ${evalId} not found`);
        }
    });
}
/**
 * Deletes all evaluations and related records with foreign keys from the database.
 * @async
 * @returns {Promise<void>}
 */
async function deleteAllEvals() {
    const db = (0, database_1.getDb)();
    await db.transaction(async (tx) => {
        await tx.delete(tables_1.evalResultsTable).run();
        await tx.delete(tables_1.evalsToPromptsTable).run();
        await tx.delete(tables_1.evalsToDatasetsTable).run();
        await tx.delete(tables_1.evalsToTagsTable).run();
        await tx.delete(tables_1.evalsTable).run();
    });
}
async function readFilters(filters, basePath = '') {
    const ret = {};
    for (const [name, filterPath] of Object.entries(filters)) {
        const globPath = path.join(basePath, filterPath);
        const filePaths = (0, glob_1.globSync)(globPath, {
            windowsPathsNoEscape: true,
        });
        for (const filePath of filePaths) {
            const finalPath = path.resolve(filePath);
            ret[name] = await (0, esm_1.importModule)(finalPath);
        }
    }
    return ret;
}
function printBorder() {
    const border = '='.repeat(constants_1.TERMINAL_MAX_WIDTH);
    logger_1.default.info(border);
}
function setupEnv(envPath) {
    if (envPath) {
        logger_1.default.info(`Loading environment variables from ${envPath}`);
        dotenv_1.default.config({ path: envPath });
    }
    else {
        dotenv_1.default.config();
    }
}
const standaloneEvalCache = new node_cache_1.default({ stdTTL: 60 * 60 * 2 }); // Cache for 2 hours
async function getStandaloneEvals({ limit = DEFAULT_QUERY_LIMIT, tag, description, } = {}) {
    const cacheKey = `standalone_evals_${limit}_${tag?.key}_${tag?.value}`;
    const cachedResult = standaloneEvalCache.get(cacheKey);
    if (cachedResult) {
        return cachedResult;
    }
    const db = (0, database_1.getDb)();
    const results = db
        .select({
        evalId: tables_1.evalsTable.id,
        description: tables_1.evalsTable.description,
        results: tables_1.evalsTable.results,
        createdAt: tables_1.evalsTable.createdAt,
        promptId: tables_1.evalsToPromptsTable.promptId,
        datasetId: tables_1.evalsToDatasetsTable.datasetId,
        tagName: tables_1.tagsTable.name,
        tagValue: tables_1.tagsTable.value,
        isRedteam: (0, drizzle_orm_1.sql) `json_extract(evals.config, '$.redteam') IS NOT NULL`.as('isRedteam'),
    })
        .from(tables_1.evalsTable)
        .leftJoin(tables_1.evalsToPromptsTable, (0, drizzle_orm_1.eq)(tables_1.evalsTable.id, tables_1.evalsToPromptsTable.evalId))
        .leftJoin(tables_1.evalsToDatasetsTable, (0, drizzle_orm_1.eq)(tables_1.evalsTable.id, tables_1.evalsToDatasetsTable.evalId))
        .leftJoin(tables_1.evalsToTagsTable, (0, drizzle_orm_1.eq)(tables_1.evalsTable.id, tables_1.evalsToTagsTable.evalId))
        .leftJoin(tables_1.tagsTable, (0, drizzle_orm_1.eq)(tables_1.evalsToTagsTable.tagId, tables_1.tagsTable.id))
        .where((0, drizzle_orm_1.and)(tag ? (0, drizzle_orm_1.and)((0, drizzle_orm_1.eq)(tables_1.tagsTable.name, tag.key), (0, drizzle_orm_1.eq)(tables_1.tagsTable.value, tag.value)) : undefined, description ? (0, drizzle_orm_1.eq)(tables_1.evalsTable.description, description) : undefined))
        .orderBy((0, drizzle_orm_1.desc)(tables_1.evalsTable.createdAt))
        .limit(limit)
        .all();
    const standaloneEvals = (await Promise.all(results.map(async (result) => {
        const { description, createdAt, evalId, promptId, datasetId, 
        // @ts-ignore
        isRedteam, } = result;
        const eval_ = await eval_1.default.findById(evalId);
        (0, invariant_1.default)(eval_, `Eval with ID ${evalId} not found`);
        const table = (await eval_.getTable()) || { body: [] };
        // @ts-ignore
        return eval_.getPrompts().map((col, index) => {
            // Compute some stats
            const pluginCounts = table.body.reduce(
            // @ts-ignore
            (acc, row) => {
                const pluginId = row.test.metadata?.pluginId;
                if (pluginId) {
                    const isPass = row.outputs[index].pass;
                    acc.pluginPassCount[pluginId] =
                        (acc.pluginPassCount[pluginId] || 0) + (isPass ? 1 : 0);
                    acc.pluginFailCount[pluginId] =
                        (acc.pluginFailCount[pluginId] || 0) + (isPass ? 0 : 1);
                }
                return acc;
            }, { pluginPassCount: {}, pluginFailCount: {} });
            return {
                evalId,
                description,
                promptId,
                datasetId,
                createdAt,
                isRedteam: isRedteam,
                ...pluginCounts,
                ...col,
            };
        });
    }))).flat();
    standaloneEvalCache.set(cacheKey, standaloneEvals);
    return standaloneEvals;
}
function providerToIdentifier(provider) {
    if ((0, types_1.isApiProvider)(provider)) {
        return provider.id();
    }
    else if ((0, types_1.isProviderOptions)(provider)) {
        return provider.id;
    }
    else if (typeof provider === 'string') {
        return provider;
    }
    return undefined;
}
function varsMatch(vars1, vars2) {
    return (0, fast_deep_equal_1.default)(vars1, vars2);
}
function resultIsForTestCase(result, testCase) {
    const providersMatch = testCase.provider
        ? providerToIdentifier(testCase.provider) === providerToIdentifier(result.provider)
        : true;
    return varsMatch(testCase.vars, result.vars) && providersMatch;
}
function renderVarsInObject(obj, vars) {
    // Renders nunjucks template strings with context variables
    if (!vars || (0, envars_1.getEnvBool)('PROMPTFOO_DISABLE_TEMPLATING')) {
        return obj;
    }
    if (typeof obj === 'string') {
        return nunjucks_1.default.renderString(obj, vars);
    }
    if (Array.isArray(obj)) {
        return obj.map((item) => renderVarsInObject(item, vars));
    }
    if (typeof obj === 'object' && obj !== null) {
        const result = {};
        for (const key in obj) {
            result[key] = renderVarsInObject(obj[key], vars);
        }
        return result;
    }
    else if (typeof obj === 'function') {
        const fn = obj;
        return renderVarsInObject(fn({ vars }));
    }
    return obj;
}
/**
 * Parses a file path or glob pattern to extract function names and file extensions.
 * Function names can be specified in the filename like this:
 * prompt.py:myFunction or prompts.js:myFunction.
 * @param basePath - The base path for file resolution.
 * @param promptPath - The path or glob pattern.
 * @returns Parsed details including function name, file extension, and directory status.
 */
function parsePathOrGlob(basePath, promptPath) {
    if (promptPath.startsWith('file://')) {
        promptPath = promptPath.slice(7);
    }
    const filePath = path.resolve(basePath, promptPath);
    let stats;
    try {
        stats = fs.statSync(filePath);
    }
    catch (err) {
        if ((0, envars_1.getEnvBool)('PROMPTFOO_STRICT_FILES')) {
            throw err;
        }
    }
    let filename = path.relative(basePath, filePath);
    let functionName;
    if (filename.includes(':')) {
        const splits = filename.split(':');
        if (splits[0] && ((0, file_1.isJavascriptFile)(splits[0]) || splits[0].endsWith('.py'))) {
            [filename, functionName] = splits;
        }
    }
    const isPathPattern = stats?.isDirectory() || /[*?{}\[\]]/.test(filePath); // glob pattern
    const safeFilename = path.relative(basePath, path.isAbsolute(filename) ? filename : path.resolve(basePath, filename));
    return {
        extension: isPathPattern ? undefined : path.parse(safeFilename).ext,
        filePath: safeFilename.startsWith(basePath) ? safeFilename : path.join(basePath, safeFilename),
        functionName,
        isPathPattern,
    };
}
/**
 * Loads content from an external file if the input is a file path, otherwise
 * returns the input as-is. Supports Nunjucks templating for file paths.
 *
 * @param filePath - The input to process. Can be a file path string starting with "file://",
 * an array of file paths, or any other type of data.
 * @returns The loaded content if the input was a file path, otherwise the original input.
 * For JSON and YAML files, the content is parsed into an object.
 * For other file types, the raw file content is returned as a string.
 *
 * @throws {Error} If the specified file does not exist.
 */
function maybeLoadFromExternalFile(filePath) {
    if (Array.isArray(filePath)) {
        return filePath.map((path) => {
            const content = maybeLoadFromExternalFile(path);
            return content;
        });
    }
    if (typeof filePath !== 'string') {
        return filePath;
    }
    if (!filePath.startsWith('file://')) {
        return filePath;
    }
    // Render the file path using Nunjucks
    const renderedFilePath = (0, templates_1.getNunjucksEngine)().renderString(filePath, {});
    const finalPath = path.resolve(cliState_1.default.basePath || '', renderedFilePath.slice('file://'.length));
    if (!fs.existsSync(finalPath)) {
        throw new Error(`File does not exist: ${finalPath}`);
    }
    const contents = fs.readFileSync(finalPath, 'utf8');
    if (finalPath.endsWith('.json')) {
        return JSON.parse(contents);
    }
    if (finalPath.endsWith('.yaml') || finalPath.endsWith('.yml')) {
        return js_yaml_1.default.load(contents);
    }
    if (finalPath.endsWith('.csv')) {
        const records = (0, sync_1.parse)(contents, { columns: true });
        // If single column, return array of values
        if (records.length > 0 && Object.keys(records[0]).length === 1) {
            return records.map((record) => Object.values(record)[0]);
        }
        return records;
    }
    return contents;
}
function isRunningUnderNpx() {
    return Boolean(process.env.npm_execpath?.includes('npx') ||
        process.execPath.includes('npx') ||
        process.env.npm_lifecycle_script?.includes('npx'));
}
