// promptfoo/dist/src/redteam/shared.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.doRedteamRun = doRedteamRun;
const chalk_1 = __importDefault(require("chalk"));
const fs = __importStar(require("fs"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const eval_1 = require("../commands/eval");
const logger_1 = __importStar(require("../logger"));
const share_1 = require("../share");
const util_1 = require("../util");
const default_1 = require("../util/config/default");
const generate_1 = require("./commands/generate");
async function doRedteamRun(options) {
    if (options.verbose) {
        (0, logger_1.setLogLevel)('debug');
    }
    if (options.logCallback) {
        (0, logger_1.setLogCallback)(options.logCallback);
    }
    let configPath = options.config || 'promptfooconfig.yaml';
    let redteamPath = options.output || 'redteam.yaml';
    if (options.liveRedteamConfig) {
        // Write liveRedteamConfig to a temporary file
        const tmpFile = path.join(os.tmpdir(), `redteam-${Date.now()}`) + '/redteam.yaml';
        fs.mkdirSync(path.dirname(tmpFile), { recursive: true });
        fs.writeFileSync(tmpFile, js_yaml_1.default.dump(options.liveRedteamConfig));
        redteamPath = tmpFile;
        // Do not use default config.
        configPath = tmpFile;
        logger_1.default.debug(`Using live config from ${tmpFile}`);
        logger_1.default.debug(`Live config: ${JSON.stringify(options.liveRedteamConfig, null, 2)}`);
    }
    // Generate new test cases
    logger_1.default.info('Generating test cases...');
    const redteamConfig = await (0, generate_1.doGenerateRedteam)({
        ...options,
        config: configPath,
        output: redteamPath,
        force: options.force,
        verbose: options.verbose,
        delay: options.delay,
        inRedteamRun: true,
        abortSignal: options.abortSignal,
    });
    // Check if redteam.yaml exists before running evaluation
    if (!redteamConfig || !fs.existsSync(redteamPath)) {
        logger_1.default.info('No test cases generated. Skipping scan.');
        return;
    }
    // Run evaluation
    logger_1.default.info('Running scan...');
    const { defaultConfig } = await (0, default_1.loadDefaultConfig)();
    const evalResult = await (0, eval_1.doEval)({
        ...options,
        config: [redteamPath],
        output: options.output ? [options.output] : undefined,
        cache: true,
        write: true,
        filterProviders: options.filterProviders,
        filterTargets: options.filterTargets,
    }, defaultConfig, redteamPath, {
        showProgressBar: true,
        abortSignal: options.abortSignal,
    });
    logger_1.default.info(chalk_1.default.green('\nRed team scan complete!'));
    const command = (0, util_1.isRunningUnderNpx)() ? 'npx promptfoo' : 'promptfoo';
    if (options.loadedFromCloud) {
        const url = await (0, share_1.createShareableUrl)(evalResult, false);
        logger_1.default.info(`View results: ${chalk_1.default.greenBright.bold(url)}`);
    }
    else {
        if (options.liveRedteamConfig) {
            logger_1.default.info(chalk_1.default.blue(`To view the results, click the ${chalk_1.default.bold('View Report')} button or run ${chalk_1.default.bold(`${command} redteam report`)} on the command line.`));
        }
        else {
            logger_1.default.info(chalk_1.default.blue(`To view the results, run ${chalk_1.default.bold(`${command} redteam report`)}`));
        }
    }
    // Clear the callback when done
    (0, logger_1.setLogCallback)(null);
    return evalResult;
}
