// promptfoo/dist/src/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TERMINAL_MAX_WIDTH = exports.DEFAULT_PORT = exports.SHARE_VIEW_BASE_URL = exports.DEFAULT_SHARE_VIEW_BASE_URL = exports.SHARE_API_BASE_URL = exports.DEFAULT_QUERY_LIMIT = exports.VERSION = void 0;
const package_json_1 = __importDefault(require("../package.json"));
const envars_1 = require("./envars");
exports.VERSION = package_json_1.default.version;
exports.DEFAULT_QUERY_LIMIT = 100;
// We use process.env in this file because it's imported by the next.js server as well.
// This is used for sharing evals.
exports.SHARE_API_BASE_URL = 
// TODO(ian): Backwards compatibility, 2024-04-01
process.env.NEXT_PUBLIC_PROMPTFOO_REMOTE_API_BASE_URL ||
    process.env.NEXT_PUBLIC_PROMPTFOO_BASE_URL ||
    process.env.PROMPTFOO_REMOTE_API_BASE_URL ||
    `https://api.promptfoo.dev`;
exports.DEFAULT_SHARE_VIEW_BASE_URL = (0, envars_1.getEnvString)('PROMPTFOO_SHARING_APP_BASE_URL', `https://app.promptfoo.dev`);
// This is used for creating shared eval links.
exports.SHARE_VIEW_BASE_URL = process.env.NEXT_PUBLIC_PROMPTFOO_BASE_URL ||
    process.env.PROMPTFOO_REMOTE_APP_BASE_URL ||
    exports.DEFAULT_SHARE_VIEW_BASE_URL;
exports.DEFAULT_PORT = Number.parseInt(process.env.API_PORT || '15500');
// Maximum width for terminal outputs.
exports.TERMINAL_MAX_WIDTH = process?.stdout?.isTTY && process?.stdout?.columns && process?.stdout?.columns > 10
    ? process?.stdout?.columns - 10
    : 120;
