// promptfoo/dist/src/commands/list.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listCommand = listCommand;
const chalk_1 = __importDefault(require("chalk"));
const logger_1 = __importDefault(require("../logger"));
const eval_1 = __importStar(require("../models/eval"));
const table_1 = require("../table");
const telemetry_1 = __importDefault(require("../telemetry"));
const util_1 = require("../util");
const createHash_1 = require("../util/createHash");
function listCommand(program) {
    const listCommand = program.command('list').description('List various resources');
    listCommand
        .command('evals')
        .description('List evaluations.')
        .option('--env-file, --env-path <path>', 'Path to .env file')
        .option('-n <limit>', 'Number of evals to display')
        .option('--ids-only', 'Show only eval IDs')
        .action(async (cmdObj) => {
        (0, util_1.setupEnv)(cmdObj.envPath);
        telemetry_1.default.record('command_used', {
            name: 'list evals',
        });
        await telemetry_1.default.send();
        const evals = await eval_1.default.getMany(Number(cmdObj.n) || undefined);
        if (cmdObj.idsOnly) {
            evals.forEach((evl) => logger_1.default.info(evl.id));
            return;
        }
        const vars = await eval_1.EvalQueries.getVarsFromEvals(evals);
        const tableData = evals.map((evl) => {
            const prompts = evl.getPrompts();
            const description = evl.config.description || '';
            return {
                'Eval ID': evl.id,
                Description: description.slice(0, 100) + (description.length > 100 ? '...' : ''),
                Prompts: prompts.map((p) => (0, createHash_1.sha256)(p.raw).slice(0, 6)).join(', ') || '',
                Vars: vars[evl.id]?.join(', ') || '',
            };
        });
        logger_1.default.info((0, table_1.wrapTable)(tableData));
        (0, util_1.printBorder)();
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show eval <id>')} to see details of a specific evaluation.`);
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show prompt <id>')} to see details of a specific prompt.`);
    });
    listCommand
        .command('prompts')
        .description('List prompts used')
        .option('--env-file, --env-path <path>', 'Path to .env file')
        .option('-n <limit>', 'Number of prompts to display')
        .option('--ids-only', 'Show only prompt IDs')
        .action(async (cmdObj) => {
        (0, util_1.setupEnv)(cmdObj.envPath);
        telemetry_1.default.record('command_used', {
            name: 'list prompts',
        });
        await telemetry_1.default.send();
        const prompts = (await (0, util_1.getPrompts)(Number(cmdObj.n) || undefined)).sort((a, b) => b.recentEvalId.localeCompare(a.recentEvalId));
        if (cmdObj.idsOnly) {
            prompts.forEach((prompt) => logger_1.default.info(prompt.id));
            return;
        }
        const tableData = prompts.map((prompt) => ({
            'Prompt ID': prompt.id.slice(0, 6),
            Raw: prompt.prompt.raw.slice(0, 100) + (prompt.prompt.raw.length > 100 ? '...' : ''),
            '# evals': prompt.count,
            'Most recent eval': prompt.recentEvalId.slice(0, 6),
        }));
        logger_1.default.info((0, table_1.wrapTable)(tableData));
        (0, util_1.printBorder)();
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show prompt <id>')} to see details of a specific prompt.`);
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show eval <id>')} to see details of a specific evaluation.`);
    });
    listCommand
        .command('datasets')
        .description('List datasets used')
        .option('--env-file, --env-path <path>', 'Path to .env file')
        .option('-n <limit>', 'Number of datasets to display')
        .option('--ids-only', 'Show only dataset IDs')
        .action(async (cmdObj) => {
        (0, util_1.setupEnv)(cmdObj.envPath);
        telemetry_1.default.record('command_used', {
            name: 'list datasets',
        });
        await telemetry_1.default.send();
        const datasets = (await (0, util_1.getTestCases)(Number(cmdObj.n) || undefined)).sort((a, b) => b.recentEvalId.localeCompare(a.recentEvalId));
        if (cmdObj.idsOnly) {
            datasets.forEach((dataset) => logger_1.default.info(dataset.id));
            return;
        }
        const tableData = datasets.map((dataset) => ({
            'Dataset ID': dataset.id.slice(0, 6),
            'Highest scoring prompt': dataset.prompts
                .sort((a, b) => (b.prompt.metrics?.score || 0) - (a.prompt.metrics?.score || 0))[0]
                .id.slice(0, 6),
            '# evals': dataset.count,
            '# prompts': dataset.prompts.length,
            'Most recent eval': dataset.recentEvalId.slice(0, 6),
        }));
        logger_1.default.info((0, table_1.wrapTable)(tableData));
        (0, util_1.printBorder)();
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show dataset <id>')} to see details of a specific dataset.`);
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show prompt <id>')} to see details of a specific prompt.`);
        logger_1.default.info(`Run ${chalk_1.default.green('promptfoo show eval <id>')} to see details of a specific evaluation.`);
    });
}
