// promptfoo/dist/src/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = exports.winstonLogger = exports.fileFormatter = exports.consoleFormatter = exports.LOG_LEVELS = exports.globalLogCallback = void 0;
exports.setLogCallback = setLogCallback;
exports.getLogLevel = getLogLevel;
exports.setLogLevel = setLogLevel;
const chalk_1 = __importDefault(require("chalk"));
const path_1 = __importDefault(require("path"));
const winston_1 = __importDefault(require("winston"));
const envars_1 = require("./envars");
exports.globalLogCallback = null;
function setLogCallback(callback) {
    exports.globalLogCallback = callback;
}
exports.LOG_LEVELS = {
    error: 0,
    warn: 1,
    info: 2,
    debug: 3,
};
exports.consoleFormatter = winston_1.default.format.printf((info) => {
    const message = info.message;
    // Call the callback if it exists
    if (exports.globalLogCallback) {
        (0, exports.globalLogCallback)(message);
    }
    if (info.level === 'error') {
        return chalk_1.default.red(message);
    }
    else if (info.level === 'warn') {
        return chalk_1.default.yellow(message);
    }
    else if (info.level === 'info') {
        return message;
    }
    else if (info.level === 'debug') {
        return chalk_1.default.cyan(message);
    }
    throw new Error(`Invalid log level: ${info.level}`);
});
exports.fileFormatter = winston_1.default.format.printf((info) => {
    const timestamp = new Date().toISOString();
    const location = info.location ? ` ${info.location}` : '';
    return `${timestamp} [${info.level.toUpperCase()}]${location}: ${info.message}`;
});
exports.winstonLogger = winston_1.default.createLogger({
    levels: exports.LOG_LEVELS,
    transports: [
        new winston_1.default.transports.Console({
            level: (0, envars_1.getEnvString)('LOG_LEVEL', 'info'),
            format: winston_1.default.format.combine(winston_1.default.format.simple(), exports.consoleFormatter),
        }),
    ],
});
if (!(0, envars_1.getEnvString)('PROMPTFOO_DISABLE_ERROR_LOG', '')) {
    exports.winstonLogger.on('data', (chunk) => {
        if (chunk.level === 'error' &&
            !exports.winstonLogger.transports.some((t) => t instanceof winston_1.default.transports.File)) {
            // Only create the errors file if there are any errors
            const fileTransport = new winston_1.default.transports.File({
                filename: path_1.default.join((0, envars_1.getEnvString)('PROMPTFOO_LOG_DIR', '.'), 'promptfoo-errors.log'),
                level: 'error',
                format: winston_1.default.format.combine(winston_1.default.format.simple(), exports.fileFormatter),
            });
            exports.winstonLogger.add(fileTransport);
            // Re-log the error that triggered this so it's written to the file
            fileTransport.write(chunk);
        }
    });
}
function getLogLevel() {
    return exports.winstonLogger.transports[0].level;
}
function setLogLevel(level) {
    if (level in exports.LOG_LEVELS) {
        exports.winstonLogger.transports[0].level = level;
    }
    else {
        throw new Error(`Invalid log level: ${level}`);
    }
}
// Wrapper enforces strict single-string argument logging
exports.logger = Object.assign({}, exports.winstonLogger, {
    error: (message) => exports.winstonLogger.error(message),
    warn: (message) => exports.winstonLogger.warn(message),
    info: (message) => exports.winstonLogger.info(message),
    debug: (message) => exports.winstonLogger.debug(message),
    add: exports.winstonLogger.add.bind(exports.winstonLogger),
    remove: exports.winstonLogger.remove.bind(exports.winstonLogger),
    transports: exports.winstonLogger.transports,
});
exports.default = exports.logger;
