// promptfoo/dist/src/providers/defaults.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDefaultCompletionProviders = setDefaultCompletionProviders;
exports.setDefaultEmbeddingProviders = setDefaultEmbeddingProviders;
exports.getDefaultProviders = getDefaultProviders;
const envars_1 = require("../envars");
const logger_1 = __importDefault(require("../logger"));
const anthropic_1 = require("./anthropic");
const azure_1 = require("./azure");
const openai_1 = require("./openai");
const vertex_1 = require("./vertex");
const vertexUtil_1 = require("./vertexUtil");
const COMPLETION_PROVIDERS = [
    'datasetGenerationProvider',
    'gradingJsonProvider',
    'gradingProvider',
    'llmRubricProvider',
    'suggestionsProvider',
    'synthesizeProvider',
];
const EMBEDDING_PROVIDERS = ['embeddingProvider'];
let defaultCompletionProvider;
let defaultEmbeddingProvider;
/**
 * This will override all of the completion type providers defined in the constant COMPLETION_PROVIDERS
 * @param provider - The provider to set as the default completion provider.
 */
async function setDefaultCompletionProviders(provider) {
    defaultCompletionProvider = provider;
}
async function setDefaultEmbeddingProviders(provider) {
    defaultEmbeddingProvider = provider;
}
async function getDefaultProviders(env) {
    const preferAnthropic = !(0, envars_1.getEnvString)('OPENAI_API_KEY') &&
        !env?.OPENAI_API_KEY &&
        ((0, envars_1.getEnvString)('ANTHROPIC_API_KEY') || env?.ANTHROPIC_API_KEY);
    const hasAzureApiKey = (0, envars_1.getEnvString)('AZURE_OPENAI_API_KEY') ||
        env?.AZURE_OPENAI_API_KEY ||
        (0, envars_1.getEnvString)('AZURE_API_KEY') ||
        env?.AZURE_API_KEY;
    const hasAzureClientCreds = ((0, envars_1.getEnvString)('AZURE_CLIENT_ID') || env?.AZURE_CLIENT_ID) &&
        ((0, envars_1.getEnvString)('AZURE_CLIENT_SECRET') || env?.AZURE_CLIENT_SECRET) &&
        ((0, envars_1.getEnvString)('AZURE_TENANT_ID') || env?.AZURE_TENANT_ID);
    const preferAzure = !(0, envars_1.getEnvString)('OPENAI_API_KEY') &&
        !env?.OPENAI_API_KEY &&
        (hasAzureApiKey || hasAzureClientCreds) &&
        ((0, envars_1.getEnvString)('AZURE_DEPLOYMENT_NAME') || env?.AZURE_DEPLOYMENT_NAME) &&
        ((0, envars_1.getEnvString)('AZURE_OPENAI_DEPLOYMENT_NAME') || env?.AZURE_OPENAI_DEPLOYMENT_NAME);
    let providers;
    if (preferAzure) {
        logger_1.default.debug('Using Azure OpenAI default providers');
        const deploymentName = (0, envars_1.getEnvString)('AZURE_OPENAI_DEPLOYMENT_NAME') || env?.AZURE_OPENAI_DEPLOYMENT_NAME;
        if (!deploymentName) {
            throw new Error('AZURE_OPENAI_DEPLOYMENT_NAME must be set when using Azure OpenAI');
        }
        const embeddingDeploymentName = (0, envars_1.getEnvString)('AZURE_OPENAI_EMBEDDING_DEPLOYMENT_NAME') ||
            env?.AZURE_OPENAI_EMBEDDING_DEPLOYMENT_NAME ||
            deploymentName;
        const azureProvider = new azure_1.AzureChatCompletionProvider(deploymentName, { env });
        const azureEmbeddingProvider = new azure_1.AzureEmbeddingProvider(embeddingDeploymentName, {
            env,
        });
        providers = {
            datasetGenerationProvider: azureProvider,
            embeddingProvider: azureEmbeddingProvider,
            gradingJsonProvider: azureProvider,
            gradingProvider: azureProvider,
            moderationProvider: openai_1.DefaultModerationProvider,
            suggestionsProvider: azureProvider,
            synthesizeProvider: azureProvider,
        };
    }
    else if (preferAnthropic) {
        logger_1.default.debug('Using Anthropic default providers');
        providers = {
            datasetGenerationProvider: anthropic_1.DefaultGradingProvider,
            embeddingProvider: openai_1.DefaultEmbeddingProvider, // TODO(ian): Voyager instead?
            gradingJsonProvider: anthropic_1.DefaultGradingJsonProvider,
            gradingProvider: anthropic_1.DefaultGradingProvider,
            llmRubricProvider: anthropic_1.DefaultLlmRubricProvider,
            moderationProvider: openai_1.DefaultModerationProvider,
            suggestionsProvider: anthropic_1.DefaultSuggestionsProvider,
            synthesizeProvider: anthropic_1.DefaultGradingJsonProvider,
        };
    }
    else if (!(0, envars_1.getEnvString)('OPENAI_API_KEY') &&
        !env?.OPENAI_API_KEY &&
        (await (0, vertexUtil_1.hasGoogleDefaultCredentials)())) {
        logger_1.default.debug('Using Google default providers');
        providers = {
            datasetGenerationProvider: vertex_1.DefaultGradingProvider,
            embeddingProvider: vertex_1.DefaultEmbeddingProvider,
            gradingJsonProvider: vertex_1.DefaultGradingProvider,
            gradingProvider: vertex_1.DefaultGradingProvider,
            moderationProvider: openai_1.DefaultModerationProvider,
            suggestionsProvider: vertex_1.DefaultGradingProvider,
            synthesizeProvider: vertex_1.DefaultGradingProvider,
        };
    }
    else {
        logger_1.default.debug('Using OpenAI default providers');
        providers = {
            datasetGenerationProvider: openai_1.DefaultGradingProvider,
            embeddingProvider: openai_1.DefaultEmbeddingProvider,
            gradingJsonProvider: openai_1.DefaultGradingJsonProvider,
            gradingProvider: openai_1.DefaultGradingProvider,
            moderationProvider: openai_1.DefaultModerationProvider,
            suggestionsProvider: openai_1.DefaultSuggestionsProvider,
            synthesizeProvider: openai_1.DefaultGradingJsonProvider,
        };
    }
    if (defaultCompletionProvider) {
        logger_1.default.debug(`Overriding default completion provider: ${defaultCompletionProvider.id()}`);
        COMPLETION_PROVIDERS.forEach((provider) => {
            providers[provider] = defaultCompletionProvider;
        });
    }
    if (defaultEmbeddingProvider) {
        EMBEDDING_PROVIDERS.forEach((provider) => {
            providers[provider] = defaultEmbeddingProvider;
        });
    }
    return providers;
}
