// promptfoo/dist/src/validators/providers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProviderSchema = exports.ProvidersSchema = exports.ProviderClassificationResponseSchema = exports.ProviderSimilarityResponseSchema = exports.ProviderEmbeddingResponseSchema = exports.ProviderResponseSchema = exports.ApiProviderSchema = exports.CallApiOptionsParamsSchema = exports.CallApiContextParamsSchema = exports.ProviderOptionsSchema = exports.ProviderEnvOverridesSchema = void 0;
const zod_1 = require("zod");
const prompts_1 = require("./prompts");
const shared_1 = require("./shared");
// Note: sync with types/providers.ts
exports.ProviderEnvOverridesSchema = zod_1.z.object({
    AI21_API_BASE_URL: zod_1.z.string().optional(),
    AI21_API_KEY: zod_1.z.string().optional(),
    ANTHROPIC_API_KEY: zod_1.z.string().optional(),
    AWS_BEDROCK_REGION: zod_1.z.string().optional(),
    AZURE_OPENAI_API_BASE_URL: zod_1.z.string().optional(),
    AZURE_OPENAI_API_HOST: zod_1.z.string().optional(),
    AZURE_OPENAI_API_KEY: zod_1.z.string().optional(),
    AZURE_API_BASE_URL: zod_1.z.string().optional(),
    AZURE_API_HOST: zod_1.z.string().optional(),
    AZURE_API_KEY: zod_1.z.string().optional(),
    AZURE_DEPLOYMENT_NAME: zod_1.z.string().optional(),
    AZURE_EMBEDDING_DEPLOYMENT_NAME: zod_1.z.string().optional(),
    AZURE_OPENAI_BASE_URL: zod_1.z.string().optional(),
    AZURE_OPENAI_DEPLOYMENT_NAME: zod_1.z.string().optional(),
    AZURE_OPENAI_EMBEDDING_DEPLOYMENT_NAME: zod_1.z.string().optional(),
    AZURE_CLIENT_SECRET: zod_1.z.string().optional(),
    AZURE_CLIENT_ID: zod_1.z.string().optional(),
    AZURE_TENANT_ID: zod_1.z.string().optional(),
    AZURE_AUTHORITY_HOST: zod_1.z.string().optional(),
    AZURE_TOKEN_SCOPE: zod_1.z.string().optional(),
    BAM_API_HOST: zod_1.z.string().optional(),
    BAM_API_KEY: zod_1.z.string().optional(),
    CLOUDFLARE_ACCOUNT_ID: zod_1.z.string().optional(),
    CLOUDFLARE_API_KEY: zod_1.z.string().optional(),
    PROMPTFOO_REMOTE_GENERATION_URL: zod_1.z.string().optional(),
    COHERE_API_KEY: zod_1.z.string().optional(),
    FAL_KEY: zod_1.z.string().optional(),
    GOOGLE_API_HOST: zod_1.z.string().optional(),
    GOOGLE_API_KEY: zod_1.z.string().optional(),
    GROQ_API_KEY: zod_1.z.string().optional(),
    LOCALAI_BASE_URL: zod_1.z.string().optional(),
    MISTRAL_API_BASE_URL: zod_1.z.string().optional(),
    MISTRAL_API_HOST: zod_1.z.string().optional(),
    MISTRAL_API_KEY: zod_1.z.string().optional(),
    OPENAI_API_BASE_URL: zod_1.z.string().optional(),
    OPENAI_API_HOST: zod_1.z.string().optional(),
    OPENAI_API_KEY: zod_1.z.string().optional(),
    OPENAI_BASE_URL: zod_1.z.string().optional(),
    OPENAI_ORGANIZATION: zod_1.z.string().optional(),
    PALM_API_HOST: zod_1.z.string().optional(),
    PALM_API_KEY: zod_1.z.string().optional(),
    REPLICATE_API_KEY: zod_1.z.string().optional(),
    REPLICATE_API_TOKEN: zod_1.z.string().optional(),
    VERTEX_API_HOST: zod_1.z.string().optional(),
    VERTEX_API_KEY: zod_1.z.string().optional(),
    VERTEX_PROJECT_ID: zod_1.z.string().optional(),
    VERTEX_PUBLISHER: zod_1.z.string().optional(),
    VERTEX_REGION: zod_1.z.string().optional(),
    WATSONX_AI_APIKEY: zod_1.z.string().optional(),
    WATSONX_AI_PROJECT_ID: zod_1.z.string().optional(),
    WATSONX_AI_BEARER_TOKEN: zod_1.z.string().optional(),
    WATSONX_AI_AUTH_TYPE: zod_1.z.string().optional(),
    PROMPTFOO_CA_CERT_PATH: zod_1.z.string().optional(),
    PROMPTFOO_INSECURE_SSL: zod_1.z.string().optional(),
});
exports.ProviderOptionsSchema = zod_1.z
    .object({
    id: zod_1.z.custom().optional(),
    label: zod_1.z.custom().optional(),
    config: zod_1.z.any().optional(),
    prompts: zod_1.z.array(zod_1.z.string()).optional(),
    transform: zod_1.z.string().optional(),
    delay: zod_1.z.number().optional(),
    env: exports.ProviderEnvOverridesSchema.optional(),
})
    .strict();
exports.CallApiContextParamsSchema = zod_1.z.object({
    fetchWithCache: zod_1.z.optional(zod_1.z.any()),
    filters: shared_1.NunjucksFilterMapSchema.optional(),
    getCache: zod_1.z.optional(zod_1.z.any()),
    logger: zod_1.z.optional(zod_1.z.any()),
    originalProvider: zod_1.z.optional(zod_1.z.any()),
    prompt: prompts_1.PromptSchema,
    vars: zod_1.z.record(zod_1.z.union([zod_1.z.string(), zod_1.z.object({})])),
});
exports.CallApiOptionsParamsSchema = zod_1.z.object({
    includeLogProbs: zod_1.z.optional(zod_1.z.boolean()),
});
const CallApiFunctionSchema = zod_1.z
    .function()
    .args(zod_1.z.string().describe('prompt'), exports.CallApiContextParamsSchema.optional(), exports.CallApiOptionsParamsSchema.optional())
    .returns(zod_1.z.promise(zod_1.z.custom()))
    .and(zod_1.z.object({ label: zod_1.z.string().optional() }));
exports.ApiProviderSchema = zod_1.z.object({
    id: zod_1.z.function().returns(zod_1.z.string()),
    callApi: zod_1.z.custom(),
    callEmbeddingApi: zod_1.z
        .function()
        .args(zod_1.z.string())
        .returns(zod_1.z.promise(zod_1.z.custom()))
        .optional(),
    callClassificationApi: zod_1.z
        .function()
        .args(zod_1.z.string())
        .returns(zod_1.z.promise(zod_1.z.custom()))
        .optional(),
    label: zod_1.z.custom().optional(),
    transform: zod_1.z.string().optional(),
    delay: zod_1.z.number().optional(),
    config: zod_1.z.any().optional(),
});
exports.ProviderResponseSchema = zod_1.z.object({
    cached: zod_1.z.boolean().optional(),
    cost: zod_1.z.number().optional(),
    error: zod_1.z.string().optional(),
    logProbs: zod_1.z.array(zod_1.z.number()).optional(),
    metadata: zod_1.z
        .object({
        redteamFinalPrompt: zod_1.z.string().optional(),
    })
        .catchall(zod_1.z.any())
        .optional(),
    output: zod_1.z.union([zod_1.z.string(), zod_1.z.any()]).optional(),
    tokenUsage: shared_1.TokenUsageSchema.optional(),
});
exports.ProviderEmbeddingResponseSchema = zod_1.z.object({
    error: zod_1.z.string().optional(),
    embedding: zod_1.z.array(zod_1.z.number()).optional(),
    tokenUsage: shared_1.TokenUsageSchema.partial().optional(),
});
exports.ProviderSimilarityResponseSchema = zod_1.z.object({
    error: zod_1.z.string().optional(),
    similarity: zod_1.z.number().optional(),
    tokenUsage: shared_1.TokenUsageSchema.partial().optional(),
});
exports.ProviderClassificationResponseSchema = zod_1.z.object({
    error: zod_1.z.string().optional(),
    classification: zod_1.z.record(zod_1.z.number()).optional(),
});
exports.ProvidersSchema = zod_1.z.union([
    zod_1.z.string(),
    CallApiFunctionSchema,
    zod_1.z.array(zod_1.z.union([
        zod_1.z.string(),
        zod_1.z.record(zod_1.z.string(), exports.ProviderOptionsSchema),
        exports.ProviderOptionsSchema,
        CallApiFunctionSchema,
    ])),
]);
exports.ProviderSchema = zod_1.z.union([zod_1.z.string(), exports.ProviderOptionsSchema, exports.ApiProviderSchema]);
// Ensure that schemas match their corresponding types
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function assert() { }
assert();
assert();
assert();
assert();
assert();
assert();
assert();
