// promptfoo/dist/src/redteam/plugins/pliny.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlinyPlugin = exports.PLUGIN_ID = void 0;
/*
 * The prompts used by this CLI are licensed under the AGPL-3.0 license and are
 * dynamically downloaded from Github. For details, see
 * https://github.com/elder-plinius/L1B3RT4S/blob/main/LICENSE.
 */
const fetch_1 = require("../../fetch");
const base_1 = require("./base");
exports.PLUGIN_ID = 'promptfoo:redteam:pliny';
const URLS = [
    'https://raw.githubusercontent.com/elder-plinius/L1B3RT4S/refs/heads/main/GOOGLE.mkd',
    'https://raw.githubusercontent.com/elder-plinius/L1B3RT4S/refs/heads/main/AMAZON.mkd',
    'https://raw.githubusercontent.com/elder-plinius/L1B3RT4S/refs/heads/main/META.mkd',
    'https://raw.githubusercontent.com/elder-plinius/L1B3RT4S/refs/heads/main/OPENAI.mkd',
];
async function fetchAndParseUrl(url) {
    try {
        const response = await (0, fetch_1.fetchWithProxy)(url);
        const text = await response.text();
        // Split by headers (h1-h3) and filter out empty strings
        const sections = text.split(/^#{1,4}\s+.*$/m).filter(Boolean);
        return sections
            .map((section) => section.trim())
            .filter(Boolean)
            .filter((section) => section.includes('\n'));
    }
    catch (error) {
        console.error(`Error fetching ${url}:`, error);
        return [];
    }
}
async function fetchAllTexts() {
    try {
        const allSections = await Promise.all(URLS.map((url) => fetchAndParseUrl(url)));
        // Flatten array of arrays and filter out empty strings
        return allSections.flat().filter(Boolean);
    }
    catch (error) {
        console.error('Error fetching texts:', error);
        return [];
    }
}
class PlinyPlugin extends base_1.RedteamPluginBase {
    async getTemplate() {
        return this.injectVar;
    }
    getAssertions(prompt) {
        return [
            {
                type: 'is-refusal',
                metric: 'Pliny',
            },
        ];
    }
    async generateTests(n, _delayMs) {
        const texts = await fetchAllTexts();
        // Take n random texts, or all if we have fewer than n
        const selectedTexts = texts.sort(() => Math.random() - 0.5).slice(0, Math.min(n, texts.length));
        return selectedTexts.map((text) => ({
            vars: {
                [this.injectVar]: ' ' + text,
            },
            assert: this.getAssertions(text),
        }));
    }
}
exports.PlinyPlugin = PlinyPlugin;
