// promptfoo/dist/src/globalConfig/cloud.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloudConfig = exports.CloudConfig = void 0;
const fetch_1 = require("../fetch");
const logger_1 = __importDefault(require("../logger"));
const globalConfig_1 = require("./globalConfig");
const API_HOST = process.env.API_HOST || 'https://api.promptfoo.app';
class CloudConfig {
    constructor() {
        const savedConfig = (0, globalConfig_1.readGlobalConfig)()?.cloud || {};
        this.config = {
            appUrl: savedConfig.appUrl || 'https://www.promptfoo.app',
            apiHost: savedConfig.apiHost || API_HOST,
            apiKey: savedConfig.apiKey,
        };
    }
    isEnabled() {
        return !!this.config.apiKey;
    }
    setApiHost(apiHost) {
        this.config.apiHost = apiHost;
        this.saveConfig();
    }
    setApiKey(apiKey) {
        this.config.apiKey = apiKey;
        this.saveConfig();
    }
    getApiKey() {
        return this.config.apiKey;
    }
    getApiHost() {
        return this.config.apiHost;
    }
    setAppUrl(appUrl) {
        this.config.appUrl = appUrl;
        this.saveConfig();
    }
    getAppUrl() {
        return this.config.appUrl;
    }
    delete() {
        (0, globalConfig_1.writeGlobalConfigPartial)({ cloud: {} });
    }
    saveConfig() {
        (0, globalConfig_1.writeGlobalConfigPartial)({ cloud: this.config });
        this.reload();
    }
    reload() {
        const savedConfig = (0, globalConfig_1.readGlobalConfig)()?.cloud || {};
        this.config = {
            appUrl: savedConfig.appUrl || 'https://www.promptfoo.app',
            apiHost: savedConfig.apiHost || API_HOST,
            apiKey: savedConfig.apiKey,
        };
    }
    async validateAndSetApiToken(token, apiHost) {
        const response = await (0, fetch_1.fetchWithProxy)(`${apiHost}/users/me`, {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        });
        if (!response.ok) {
            throw new Error('Failed to validate API token: ' + response.statusText);
        }
        logger_1.default.info('You are logged in successfully.');
        const { user, organization, app } = await response.json();
        this.setApiKey(token);
        this.setApiHost(apiHost);
        this.setAppUrl(app.url);
        logger_1.default.info('Logged in as:');
        logger_1.default.info(`User: ${user.email}`);
        logger_1.default.info(`Organization: ${organization.name}`);
        logger_1.default.info(`Access the app at ${app.url}`);
        return {
            user,
            organization,
            app,
        };
    }
}
exports.CloudConfig = CloudConfig;
// Export a singleton instance
exports.cloudConfig = new CloudConfig();
